ui/src/BaseEditor.ts

```diff
@@ -0,0 +1,123 @@
+
+    public async showEditor(context: ContextT, sizeLimit?: number /* in Megabytes */): Promise<void> {
+        const fileName: string = await this.getFilename(context);
+
+        this.appendLineToOutput(localize('opening', 'Opening "{0}"...', fileName));
+        if (sizeLimit !== undefined) {
+            const size: number = await this.getSize(context);
+            if (size > sizeLimit) {
+                const message: string = localize('tooLargeError', '"{0}" is too large to download.', fileName);
+                throw new Error(message);
+            }
+        }
+
+        const localFilePath: string = await createTemporaryFile(fileName);
+        const document: vscode.TextDocument = await vscode.workspace.openTextDocument(localFilePath);
+        this.fileMap[localFilePath] = [document, context];
+        const data: string = await this.getData(context);
+        const textEditor: vscode.TextEditor = await vscode.window.showTextDocument(document);
+        await this.updateEditor(data, textEditor);
+    }
+
+    public async updateMatchingContext(doc: vscode.Uri): Promise<void> {
+        const filePath: string | undefined = Object.keys(this.fileMap).find((fsPath: string) => path.relative(doc.fsPath, fsPath) === '');
+        if (filePath) {
+            const [textDocument, context]: [vscode.TextDocument, ContextT] = this.fileMap[filePath];
+            await this.updateRemote(context, textDocument);
+        }
+    }
+
+    public async dispose(): Promise<void> {
+        Object.keys(this.fileMap).forEach(async (key: string) => await fse.remove(path.dirname(key)));
+    }
+
+    public async onDidSaveTextDocument(globalState: vscode.Memento, doc: vscode.TextDocument): Promise<void> {
+        const filePath: string | undefined = Object.keys(this.fileMap).find((fsPath: string) => path.relative(doc.uri.fsPath, fsPath) === '');
+        if (!this.ignoreSave && filePath) {
+            const context: ContextT = this.fileMap[filePath][1];
+            const showSaveWarning: boolean | undefined = vscode.workspace.getConfiguration().get(this.showSavePromptKey);
+
+            if (showSaveWarning) {
+                const message: string = await this.getSaveConfirmationText(context);
+                const result: vscode.MessageItem | undefined = await vscode.window.showWarningMessage(message, DialogResponses.upload, DialogResponses.dontWarn, DialogResponses.dontUpload);
+                if (result === DialogResponses.dontWarn) {
+                    await vscode.workspace.getConfiguration().update(this.showSavePromptKey, false, vscode.ConfigurationTarget.Global);
+                    await globalState.update(this.showSavePromptKey, true);
+                } else if (result === DialogResponses.dontUpload) {
+                    throw new UserCancelledError();
+                }
+            }
+            await this.updateRemote(context, doc);
+        }
+    }
+
+    protected appendToOutput(value: string): void {

```

Comment:
I would remove this function. As discussed, we should always use appendLine because you never know what the state of the outputChannel is